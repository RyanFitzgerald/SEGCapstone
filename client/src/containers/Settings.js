import React from 'react';
import { Switch, Route, Redirect } from 'react-router-dom';
import * as api from '../api';
import arraySort from 'array-sort';

// Import setting components
import Options from '../components/settings/Options';
import Submenu from '../components/settings/Submenu';
import Add from '../components/users/Add';
import Edit from '../components/users/Edit';
import Directory from '../components/users/Directory';
import View from '../components/users/View';
import { ToastContainer, toast } from 'react-toastify';
import 'react-toastify/dist/ReactToastify.min.css';

class Settings extends React.Component {
  constructor() {
    super();

    // Bind functions
    this.setActiveSubtab = this.setActiveSubtab.bind(this);
    this.getUsers = this.getUsers.bind(this);
    this.getRoles = this.getRoles.bind(this);
    this.addToUsers = this.addToUsers.bind(this);
    this.renderError = this.renderError.bind(this);
    this.removeFromUsers = this.removeFromUsers.bind(this);
    this.updateUsers = this.updateUsers.bind(this);
    this.sortByKey = this.sortByKey.bind(this);
    
    // State
    this.state = {
      activeSubtab: 1,
      users: null,
      roles: null,
      sort: {
        name: null,
        email: null
      }
    };
  }
  
  componentDidMount() {
    // Set Page Title
    document.title = 'Users Overview | Renovaction';

    // Update tab
    this.props.setActiveTab(5);

    // Get users
    this.getUsers({search: false});

    // Get roles
    this.getRoles();
  }
  
  setActiveSubtab(tab) {
    this.setState({activeSubtab: tab});
  }

  getRoles() {
    api.getRoles({access_token: JSON.parse(sessionStorage.getItem('user')).access_token}).then(result => {
      const roles = result;
      this.setState({ roles });
    });
  }

  getUsers(query) {
    // Append access token
    query.access_token = JSON.parse(sessionStorage.getItem('user')).access_token;

    api.getUsers(query).then(result => {
      const users = result;
      this.setState({ users });
    });
  }

  addToUsers(user) {
    const users = [...this.state.users];
    users[this.state.users.length] = user;
    this.setState({ users });
  }

  removeFromUsers(id) {
    const users = [...this.state.users];
    const updated = users.filter(el => {
      return el._id !== id;
    });
    this.setState({ users: updated });
  }
  
  updateUsers(user) {
    const users = [...this.state.users];
    const key = Object.keys(users).find(key => users[key]._id === user._id);
    users[key] = user;
    this.setState({ users });
  }

  addNotification(message, type) {
    toast(message, { type });
  }

  sortByKey(array, key) {
    let asc = 'asc';
    let desc = 'desc'
    let sortOrder = {name: null, email: null, role: null};
    const arr = Object.keys(array).map((k) => array[k]);
    let sortedArray = [];

    if (key === 'name') {
      if (this.state.sort.name === asc) {
        sortOrder.name = desc;
        sortedArray = arraySort(arr, key, {reverse: true});
      } else {
        sortOrder.name = asc;
        sortedArray = arraySort(arr, key);
      }
    } else if (key === 'email') {
      if (this.state.sort.email === asc) {
        sortOrder.email = desc;
        sortedArray = arraySort(arr, key, {reverse: true});
      } else {
        sortOrder.email = asc;
        sortedArray = arraySort(arr, key);;
      }
    } else if (key === 'role.name') {
      if (this.state.sort.role === asc) {
        sortOrder.role = desc;
        sortedArray = arraySort(arr, key, {reverse: true});
      } else {
        sortOrder.role = asc;
        sortedArray = arraySort(arr, key);;
      }
    }
    
    this.setState({users: sortedArray, sort: sortOrder});
  }

  renderError(formError) {
    if (!formError) return;
    return (
      <div className="flash flash--warn">
        <p>{formError}</p>
      </div>
    );
  }
  
  render() {
    const level = JSON.parse(sessionStorage.getItem('user')).role.level;

    return (
      <div>
        <ToastContainer 
          position="top-right"
          type="success"
          autoClose={3000}
          hideProgressBar
          newestOnTop
          closeOnClick
          pauseOnHover
        />
        <Submenu activeSubtab={this.state.activeSubtab} level={this.props.level} checkLevel={this.props.checkLevel}/>

        <Switch>
          <Route exact path="/settings" render={() => (
            (level < 2) ? (
              <Redirect to='/'/>
            ) : (
              <Options setActiveSubtab={this.setActiveSubtab}/>
            )
          )}/>
          <Route path="/settings/users/add" render={() => (
            (level < 2) ? (
              <Redirect to='/'/>
            ) : (
              <Add setActiveSubtab={this.setActiveSubtab} roles={this.state.roles} addNotification={this.addNotification} renderError={this.renderError} addToUsers={this.addToUsers}/>
            )
          )}/>
          <Route path="/settings/users/:id/edit" render={(location) => (
            (level < 2) ? (
              <Redirect to='/'/>
            ) : (
              <Edit setActiveSubtab={this.setActiveSubtab} addNotification={this.addNotification} renderError={this.renderError} location={location} updateUsers={this.updateUsers}/>
            )
          )}/>
          <Route path="/settings/users/:id" render={(location) => (
            (level < 2) ? (
              <Redirect to='/'/>
            ) : (
              <View setActiveSubtab={this.setActiveSubtab} addNotification={this.addNotification} location={location} removeFromUsers={this.removeFromUsers}/>
            )
          )}/> 
          <Route path="/settings/users" render={() => (
            (level < 2) ? (
              <Redirect to='/'/>
            ) : (
              <Directory setActiveSubtab={this.setActiveSubtab} roles={this.state.roles} users={this.state.users} getUsers={this.getUsers} sortByKey={this.sortByKey}/>
            )
          )}/>
        </Switch>
      </div>
    );
  }
}

  export default Settings;