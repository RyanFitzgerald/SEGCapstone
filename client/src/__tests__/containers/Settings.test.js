import React from 'react';
import Settings from '../../Containers/Settings';

describe('<Settings />', () => {
  beforeAll(() => {
    // Add necessary info to user
    const user = {
      role: {
        level: 1
      }
    };
    localStorage.setItem('user', JSON.stringify(user));
  });

  it('should render correctly', () => {
    // Return empty objects on requests
    mock.onAny().reply(200, {});

    // Create mock functions / spies
    const setActiveTabMock = jest.fn();
    const getUsersSpy = jest.spyOn(Settings.prototype, 'getUsers');
    const getRolesSpy = jest.spyOn(Settings.prototype, 'getRoles');

    // Shallow render it
    const wrapper = shallow(
      <Settings setActiveTab={setActiveTabMock}/>
    );

    // Compare to old snapshot
    expect(shallowToJson(wrapper)).toMatchSnapshot();

    // Check set active tab is called and with correct value
    expect(setActiveTabMock).toHaveBeenCalledTimes(1);
    expect(setActiveTabMock.mock.calls[0][0]).toBe(5);

    // Check if component mounting methods were called
    expect(getUsersSpy).toHaveBeenCalled();
    expect(getRolesSpy).toHaveBeenCalled();
  });
});

