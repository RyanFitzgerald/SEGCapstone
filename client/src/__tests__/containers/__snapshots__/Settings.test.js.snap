// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Settings /> should render correctly 1`] = `
<div>
  <ToastContainer
    autoClose={3000}
    bodyClassName=""
    className={null}
    closeButton={<DefaultCloseButton />}
    closeOnClick={true}
    hideProgressBar={true}
    newestOnTop={true}
    pauseOnHover={true}
    position="top-right"
    progressClassName=""
    style={null}
    toastClassName=""
    transition={[Function]}
    type="success"
  />
  <Submenu
    activeSubtab={1}
  />
  <Switch>
    <Route
      exact={true}
      path="/settings"
      render={[Function]}
    />
    <Route
      path="/settings/users/add"
      render={[Function]}
    />
    <Route
      path="/settings/users/:id/edit"
      render={[Function]}
    />
    <Route
      path="/settings/users/:id"
      render={[Function]}
    />
    <Route
      path="/settings/users"
      render={[Function]}
    />
  </Switch>
</div>
`;
